osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs

```diff
@@ -20,30 +21,38 @@ public class ManiaModConstant : Mod, IApplicableToBeatmap
 
         public void ApplyToBeatmap(IBeatmap beatmap)
         {
-            // Get all difficulty points
-            var difficultyPoints = beatmap.ControlPointInfo.DifficultyPoints;
-            // Get all timing points
-            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
             // Get the initial BPM
-            var initialBPM = timingPoints[0].BPM;
+            var initialBPM = beatmap.ControlPointInfo.TimingPoints[0].BPM;
 
-            foreach (var difficultyPoint in difficultyPoints)
+            foreach (var group in beatmap.ControlPointInfo.Groups)
             {
-                // Set this difficulty point's speed multiplier to 1
-                difficultyPoint.SpeedMultiplier = 1;
-            }
+                // Get the timing control point of this group, if any
+                var timingControlPoints = group.ControlPoints.OfType<TimingControlPoint>().ToArray();
 
-            // Correct changes in speed caused by timing points by adding an associated difficulty point
-            foreach (var timingPoint in timingPoints)
-            {
-                // Create a new difficulty point to counteract timing point
-                DifficultyControlPoint diffControlPoint = new DifficultyControlPoint
+                // Get the difficulty control point of this group, if any
+                var difficultyControlPoints = group.ControlPoints.OfType<DifficultyControlPoint>().ToArray();
+
+                // If there is a timing control point
+                if (timingControlPoints.Any())
+                {
+                    var timingControlPoint = timingControlPoints[0];
+
+                    DifficultyControlPoint diffControlPoint = new DifficultyControlPoint();
+
+                    diffControlPoint.SpeedMultiplierBindable.MinValue = double.MinValue; // Uncapped minimum value
```

Comment:
Yeah I'm not sure this should be allowed to be negative (or even zero). I'd agree with

```suggestion
                    diffControlPoint.SpeedMultiplierBindable.MinValue = Precision.DOUBLE_EPSILON; // Uncapped minimum value
```

Reply:
Alright, I'll fix this as well as the precision.